src/lib/cli.ts

```diff
@@ -0,0 +1,149 @@
+/**
+ * @fileoverview Main CLI object, it reads the options from cli.
+ */
+
+/*
+ * The CLI object should *not* call process.exit() directly. It should only return
+ * exit codes. This allows other programs to use the CLI object and still control
+ * when the program exits.
+ */
+
+import * as path from 'path';
+
+import { debug as d } from './utils/debug';
+import * as globby from 'globby';
+import chalk from 'chalk';
+
+const debug: debug.IDebugger = d(__filename);
+
+import { options } from './cli/options';
+import { CLIOptions, IMDFile, ILink } from './types';
+import { MDFile } from './mdfile';
+
+const getMDFiles = async (directory): Promise<IMDFile[]> => {
+    const filesPath = await globby(['**/*.md', '!node_modules', '!**/node_modules'], { cwd: directory });
+
+    return filesPath.map((relativePath) => {
+        const file = new MDFile(directory, relativePath);
+
+        return file;
+    });
+};
+
+const validateLinks = async (mdFiles: IMDFile[]): Promise<void> => {
+    for (const mdFile of mdFiles) {
+        // Validate file by file to prevent 429 and socket hang up errors.
+        await mdFile.validateLinks();
```

Comment:
Are we validating all links from all files at the same time? If so maybe we should look into parallelizing something? Won't one-by-one take a long time? Maybe parallelize the ones that are for the file system?

Reply:
Validate relative and internal links is fast, I think we don't need to improve anything about that. For absolute links we can't validate all at the same time, or we get network or "too many request" errors, That is why I am validating file by file.

Reply:
> validating file by file.

Validating one by one you mean, right?
How long does it take to validate all the links with the current code? That will allow us to better decide if we should parallelize or not.

Reply:
Aprox. 127 seconds to validate `hint` repo.

Reply:
How many links are there? 2 minutes is quite a lot IMO 😖

Reply:
2152 "absolute" links.

139 internal(links to the same file) + relative. This takes just 0.6 seconds, it is not worth doing anything to improve it.

I already have a cache for requests and first of all I request using `head`. What else can we do to improve that time?

Maybe we can try to scan more than one file at the same time, but we are increasing the chance to get errors because of the number of petitions.

Anyways, I'm going to try to analyze all the files at the same time, because now we have retries.

Reply:
Playing with the retries and the cache, I can analyze all the files at the same time, but this is not very reliable. Maybe I can try to parallelize just a few files. Does it sound good?

Reply:
Where does the script spend most of the time?
I think that analyzing multiple files at the same time looks like a good first step. Do you request the external urls in parallel for a single file or is always one by one?

Reply:
I was requesting all the link in one file at the same time, but file by file.
I'm going to increase the number of files at the same time.

Reply:
Doing chunks of 10 files, the analysis takes 40 seconds when everything is valid.

Reply:
Nice! Have you tried with more files? Is that the sweet spot?

Reply:
No, I can try with more, but it is going to depend on the files and the links on them, what works today, tomorrow can fail.

Reply:
Please give it a try. No need to be super exact (i.e.: increase by 5 each time). Any savings are welcome :)

Reply:
Take a look to the new solution. Instead of analyzing group of 10 files, I'm analyzing up to 10 files at the same time all the time.

With this solution, the time with a limit of 10 files and 20 files is more or less the same, so I prefer to keep 10.

btw, the time is more or less 22s in my computer.